python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,518 @@
+    """Print individual elapses and their average."""
+
+    def s_to_us(seconds):
+        """Converts seconds to microseconds (rounded)."""
+        return round(seconds * 1e6)
+
+    for i, elapse in enumerate(elapses):
+        print(f"{desc} #{i} elapse={s_to_us(elapse)} us")
+    total = sum(elapses)
+    avg = total / len(elapses)
+    print(f"{desc} avg: {s_to_us(avg)} us")
+    total -= elapses[0]
+    avg = total / (len(elapses) - 1)
+    print(f"{desc} avg w/o 1st iter: {s_to_us(avg)} us")
+
+
+def get_adag_ddp_weights_per_device(
+    weights: List[List[Tuple[torch.Tensor]]], num_actors: int
+):
+    """
+    Check that the model is consistent across all ranks after each iteration
+    of training. Deduplicate the weights after the checks.
+
+    Args:
+        weights: Weights from running aDAG DDP for a number of iterations.
+            `weights[i]` is the weights across all actors for the ith iteration.
+    """
+    weights_per_device = []
+    for single_iter_weights in weights:
+        per_device_len = len(single_iter_weights[0])
+        for i in range(1, num_actors):
+            for j in range(per_device_len):
+                assert torch.equal(single_iter_weights[0][j], single_iter_weights[i][j])
+        weights_per_device.append(single_iter_weights[0])
+    return weights_per_device
+
+
+def detach_all(tensors_across_iters: List[List[torch.Tensor]]):
+    """
+    Detach all tensors in order to pass tensors through `torch.multiprocessing.spawn`.
+    """
+    all_iters_detached = []
+    for single_iter_tensors in tensors_across_iters:
+        detached = []
+        for tensor in single_iter_tensors:
+            detached.append(tensor.detach())
+        all_iters_detached.append(detached)
+    return all_iters_detached
+
+
+def compare_weights(W1, W2, desc, allow_error=False):
+    """
+    Compare the weights after each iteration across different training approaches.
+
+    Args:
+        W1: Weights after each iteration from one approach.
+        W2: Weights after each iteration from the other approach.
+        desc: Description of approaches
+        allow_error: Whether small errors are allowed.
+    """
+    assert len(W1) == len(W2)
+    # w1, w2 are weights after a single iteration for the 1st and 2nd approaches,
+    # respectively.
+    max_diff = 0
+    for w1, w2 in zip(W1, W2):
+        assert len(w1) == len(w2)
+        # t1, t2 are weights of a single layer.
+        for t1, t2 in zip(w1, w2):
+            t1 = t1.to("cpu")
+            t2 = t2.to("cpu")
+            if not allow_error:
+                assert torch.allclose(
+                    t1, t2
+                ), f"{desc} max diff: {torch.max(torch.abs(t1 - t2).flatten())}"
+            elif not torch.allclose(t1, t2):
+                max_diff = max(max_diff, torch.max(torch.abs(t1 - t2).flatten()))
+
+    if max_diff != 0:
+        print(f"{desc} max diff: {max_diff}")
+
+
+def check_correctness(adag_ddp_weights, config: Config) -> None:
```

Comment:
You should append the logic of `check_correctness()` in `main()` directly. A more clean way is something as follows:

```python
adag_weights = run_adag()
torch_weights = run_torch()
compare_weights(adag_weights, torch_weights)
torch_dist_weights = run_torch_dist()
compare_weights(adag_weights, torch_weights)
```

Reply:
It seems not necessary to pass `adag_ddp_weights` when running something else rather than `adag`.

Reply:
Returned weights from `run_torch_ddp` instead of comparing inside the function.